dnsapi/dns_opnsense.sh

```diff
@@ -0,0 +1,265 @@
+  fi
+
+  return 0
+}
+
+####################  Private functions below ##################################
+#_acme-challenge.www.domain.com
+#returns
+# _domainid=domid
+#_domain=domain.com
+_get_root() {
+  domain=$1
+  i=2
+  p=1
+  if _opns_rest "GET" "/domain/get"; then
+    _domain_response="$response"
+  else
+    return 1
+  fi
+
+  while true; do
+    h=$(printf "%s" "$domain" | cut -d . -f $i-100)
+    if [ -z "$h" ]; then
+      #not valid
+      return 1
+    fi
+    _debug h "$h"
+    id=$(echo "$_domain_response" | _egrep_o "\"[^\"]*\":{\"enabled\":\"1\",\"type\":{\"master\":{\"value\":\"master\",\"selected\":1},\"slave\":{\"value\":\"slave\",\"selected\":0}},\"masterip\":\"[^\"]*\",\"domainname\":\"${h}\"" | cut -d ':' -f 1 | cut -d '"' -f 2)
+
+    if [ -n "$id" ]; then
+      _debug id "$id"
+      _host=$(printf "%s" "$domain" | cut -d . -f 1-$p)
+      _domain="${h}"
+      _domainid="${id}"
+      return 0
+    fi
+    p=$i
+    i=$(_math $i + 1)
+  done
+  _debug "$domain not found"
+
+  return 1
+}
+
+_opns_rest() {
+  method=$1
+  ep=$2
+  data=$3
+  #Percent encode user and token
+  key=$(echo "$OPNs_Key" | tr -d "\n\r" | _url_encode)
+  token=$(echo "$OPNs_Token" | tr -d "\n\r" | _url_encode)
+
+  opnsense_url="https://${key}:${token}@${OPNs_Host}:${OPNs_Port}/api/bind${ep}"
+  export _H1="Content-Type: application/json"
+  if [ ! "$method" = "GET" ]; then
+    _debug data "$data"
+    export _H1="Content-Type: application/json"
+    response="$(_post "$data" "$opnsense_url" "" "$method")"
+  else
+    export _H1=""
+    response="$(_get "$opnsense_url")"
+  fi
+
+  if [ "$?" != "0" ]; then
+    _err "error $ep"
+    return 1
+  fi
+  _debug2 response "$response"
+
+  return 0
+}
+
+_build_record_string() {
+  _record_string="{\"record\":{\"enabled\":\"1\",\"domain\":\"$1\",\"name\":\"$2\",\"type\":\"TXT\",\"value\":\"$3\"}}"
+}
+
+_existingchallenge() {
+  if _opns_rest "GET" "/record/searchRecord"; then
+    _record_response="$response"
+  else
+    return 1
+  fi
+  _uuid=""
+  _uuid=$(echo "$_record_response" | _egrep_o "\"uuid\":\"[^\"]*\",\"enabled\":\"[01]\",\"domain\":\"$1\",\"name\":\"$2\",\"type\":\"TXT\",\"value\":\"$3\"" | cut -d ':' -f 2 | cut -d '"' -f 2)
+
+  if [ -n "$_uuid" ]; then
+    _debug uuid "$_uuid"
+    return 0
+  fi
+  _debug "${2}.$1{1} record not found"
+
+  return 1
+}
+
+_opns_check_auth() {
+  OPNs_Host="${OPNs_Host:-$(_readaccountconf_mutable OPNs_Host)}"
+  OPNs_Port="${OPNs_Port:-$(_readaccountconf_mutable OPNs_Port)}"
+  OPNs_Key="${OPNs_Key:-$(_readaccountconf_mutable OPNs_Key)}"
+  OPNs_Token="${OPNs_Token:-$(_readaccountconf_mutable OPNs_Token)}"
+  OPNs_Api_Insecure="${OPNs_Api_Insecure:-$(_readaccountconf_mutable OPNs_Api_Insecure)}"
+
+  if [ -z "$OPNs_Host" ]; then
+    OPNs_Host="localhost"
+    _err "You don't specify OPNsense address."
+    return 1
+  else
+    _saveaccountconf_mutable OPNs_Host "$OPNs_Host"
+  fi
+
+  if [ -z "$OPNs_Port" ]; then
+    OPNs_Port="443"
```

Comment:
when `_opns_check_auth()`  is called from the `dns_opnsense_add()` function,  OPNs_Port is empty, and not saved.

but when it's called from `dns_opnsense_rm()` function, the `OPNs_Port` is 443, not empty.  it's then saved.

Reply:
The `if...else...fi` statement at lines 228 - 231 should probably read as follows:

```
  if [ -z "$OPNs_Port" ]; then
    _info 'OPNSense port not specified. Defaulting to using port 443'
    OPNs_Port="443"
  fi
   _saveaccountconf_mutable OPNs_Port "$OPNs_Port"
```

But I don't have access to make or propose this edit.

Reply:
look at the comment 

> don't save the value to account.conf if it has a default value.
> for example, if OPNs_Api_Insecure is not set or `0`, don't save it.
> OPNs_Port also.

That is why i changed it to if...else...fi.
I have to change a little bit. Maybe I'll use different variables for default values, I'll do it soon.  Had a lot to do past weeks.

Reply:
In that case, the easiest fix is to

1) Add this line: `OPNs_DefaultPort=443` somewhere sensible near the top.

2) Change the comment at line 7 to.read `...(optional, defaults to 443 if unset)`, for clarity.

3) Replace `$OPNs_Port` by `${OPNs_Port:-$OPNs_DefaultPort}` in line 171

4) Replace the code at lines 228-231 by:

```
  if [ -z "$OPNs_Port" ]; then
    _info 'OPNSense port not specified. Defaulting to using port 443'
 else
    _saveaccountconf_mutable OPNs_Port "$OPNs_Port"
  fi
 ```

The effect is - if a port is specified it's saved. If not, OPNs_Port remains unset/zero,so it isn't saved as Neilpang points out. And if it's null, then the value of `$OPNs_DefaultPort` is substituted at call-time.

Reply:
Thanks for the tips. 

That's much cleaner then my thought about it.

Reply:
Actually from Neilpang's comment on my own recent PR I should also suggest you replace `grep -q` with `grep ... >/dev/null`. Sorry!!